qe/drivers.py

```diff
@@ -266,6 +267,296 @@ def is_connected(det_i: Determinant, det_j: Determinant) -> Tuple[int, int]:
+
+        # If this -> no double excitation of |D> will produce |J> satisfying constraint |T>
+        if len(constraint_orbitals_occupied) == 0 or len(nonconstrained_orbitals_occupied) > 2:
+            print(
+                f"No double excitations generated by the inputted |Determinant|: {det} satisfy given constraint: {constraint}"
+            )
+            return None
+
+        # Create list of possible `particles` s.to constraint
+        if len(constraint_orbitals_occupied) == 1:
+            # (One constraint orbital occupied) e.g., a1 \in |D_a> -> A same-spin (aa) double to (x_a, y_a) \in haa to (a2, a3) is necessary
+            # No same-spin (bb) or opposite-spin (ab) excitations will satisfy constraint
+            # New particles -> a2, a3
+            a_unocc_1, a_unocc_2 = (
+                (set(det_a) | set(constraint)) - (set(det_a) & set(constraint))
+            ) & set(
+                constraint
+            )  # This set will have length 2; unpack
+            paa.append((a_unocc_1, a_unocc_2))
+
+        elif len(constraint_orbitals_occupied) == 2:
+            # (Two constraint orbitals occupied) e.g., a1, a2 \in |D_a> ->
+            #   A same-spin (aa) double (x_a, y_a) \in haa to (z_a, a_unocc), where z_a\not\in|D_a>, and z_a < a1 (if excited to z_a > a1, constraint ruined!)
+            (a_unocc,) = ((set(det_a) | set(constraint)) - (set(det_a) & set(constraint))) & set(
+                constraint
+            )  # The 1 unoccupied constraint orbital
+            det_unocc_a_orbs = self.all_orbs - set(det_a)  # Unocc orbitals in |D_a>
+            for z_a in takewhile(lambda x: x < a1, det_unocc_a_orbs):
+                # z < a_unocc trivially, no need to check they are distinct
+                paa.append((z_a, a_unocc))
+            #   No same spin (bb) excitations will satisfy constraint
+            #   An oppopsite spin (ab) double (x_a, y_b) \in \hab to (a_unocc, z_b), where z\not\in|D_b>
+            det_unocc_b_orbs = set(range(self.n_orb)) - set(det_b)  # Unocc orbitals in |D_b>
+            for _, z_b in enumerate(det_unocc_b_orbs):
+                pab.append((a_unocc, z_b))
+
+        elif len(constraint_orbitals_occupied) == 3:
+            # a1, a2, a3 \in |D_a> -> |D> satisfies constraint! ->
+            #   Any same-spin (aa) double (x_a, y_a) \in haa to (w_a, z_a), where w_a, z_a < a1
+            det_unocc_a_orbs = self.all_orbs - set(det_a)
+            for w_a in takewhile(lambda x: x < a1, det_unocc_a_orbs):
+                for z_a in takewhile(lambda x: x < a1, det_unocc_a_orbs):
+                    if w_a != z_a:
+                        paa.append((w_a, z_a))
+            #   Any same-spin (bb) double (x_b, y_b) \in hbb to (w_b, z_b)
+            det_unocc_b_orbs = self.all_orbs - set(det_b)  # Unocc orbitals in |D_a>
+            for _, w_b in enumerate(det_unocc_b_orbs):
+                for _, z_b in enumerate(det_unocc_b_orbs):
```

Comment:
No need for enumerate themn

Reply:
This is needed, they're not iterables

Reply:
https://docs.python.org/3/library/functions.html#enumerate

> Return an enumerate object. iterable must be a sequence, an [iterator](https://docs.python.org/3/glossary.html#term-iterator), or some other object which supports iteration. The [__next__()](https://docs.python.org/3/library/stdtypes.html#iterator.__next__) method of the iterator returned by [enumerate()](https://docs.python.org/3/library/functions.html#enumerate) returns a tuple containing a count (from start which defaults to 0) and the values obtained from iterating over iterable.

They are iterable by the definition of the spec. 

Please don't use `enumerate` who is just here to return a index and the value, is the first thing you do is dropping the index.

Reply:
Ahh I didn't realize, sorry!! Fixed now.